Add support for `PUT /api/v1/metadata/online`

MVrachev opened the pull request:
# Description
<!--- What is the PR about? -->

Add support for `PUT /api/v1/metadata` which will force a new metadata version
of a given set of online roles.


<!--- Please reference the issue(s) this PR fixes. -->
Fixes #399


# Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)


# Additional requirements
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature


# Code of Conduct
By submitting this PR, you agree to follow our [Code of Conduct](CODE_OF_CONDUCT.rst).

- [x] I agree to follow this project's Code of Conduct

MVrachev commented:
It's best to test this together with https://github.com/repository-service-tuf/repository-service-tuf-api/pull/513

MVrachev commented:
Before merging this pr, lets merge:  #514.

MVrachev commented:
@kairoaraujo it's ready for a review.

kairoaraujo commented on repository_service_tuf_api/api/metadata.py:
```diff
@@ -31,6 +31,25 @@ def post(payload: metadata.MetadataPostPayload):
 
 
+@router.put(
+    "/",
```
I think it should be a specific endpoint "/api/v1/metadata/<name>"
Also, I would use the following behavior. 
No rolename = bump all online roles.

kairoaraujo commented on repository_service_tuf_api/metadata.py:
```diff
@@ -85,6 +85,63 @@ def post_metadata(payload: MetadataPostPayload) -> MetadataPostResponse:
+                ),
+            },
+        )
+
```
We should verify that it is an online role. See my comment on https://github.com/repository-service-tuf/repository-service-tuf-worker/pull/435 for more detail.

kairoaraujo commented on repository_service_tuf_api/api/metadata.py:
```diff
@@ -31,6 +31,25 @@ def post(payload: metadata.MetadataPostPayload):
 
 
+@router.put(
+    "/",
```
I think it should be a specific endpoint "/api/v1/metadata/"
Also, I would use the following behavior.
No rolename = bump all online roles.

MVrachev replied:
It is already using `api/v1/metadata`:
<img width="1619" alt="image" src="https://github.com/repository-service-tuf/repository-service-tuf-api/assets/16246778/859b8aac-e829-482b-b74e-623867befe2a">

I will address the second comment though.

kairoaraujo replied:
Sorry, what I mean is to use a more specific endpoint `/api/v1/metadata/<choose a name>`
I see the `/api/v1/metadata` for more generic actions like Update Metadata.

As this description says, "Force a new version of online metadata role(s)." it should have a more specific endpoint
i.e.: `/api/v1/metadata/online`. Following our past changes, it is an asynchronous task that could be a POST.

MVrachev replied:
After a discussion we agreed it will be `POST /api/v1/metadata/online/`

MVrachev commented:
@kairoaraujo I did as you suggested if payload is empty I made sure force update on all roles.
I also added a check in `Roles.online_roles` to whether `Targets` is an online role.
If it's not we don't return it in this list.

ivanayov commented on docs/swagger.json:
```diff
@@ -143,6 +143,49 @@
+                    "Metadata"
+                ],
+                "summary": "Force a new version of online metadata role(s).",
+                "description": "Force a new version of online metadata role(s). If the roles list is empty all roles will be updated. The new metadata version(s) will have extended expiration which will equal to: today + ROLE_NAME_EXPIRATION number of days, where ROLE_NAME_EXPIRATION is a tuf repository setting. Note: depending on which metadata role you want to update other online roles will lik ...
```
What will happen if `datetime.now() + <ROLE_NAME_EXPIRATION number of days>` is less than the current expiration? I think we should either make it impossible or define it more strictly here (like extended or redused - practically it can also be the same, so we can just say "updated").

ivanayov replied:
What I mean is that this description creates expectations that don't strictly match the functionality.

MVrachev replied:
The whole idea is to be able to redefine current expiration.
For context: the idea of this feature came to life after I implemented `PUT /api/v1/config`.
With that API endpoint, you can change the expiration policy for each role.
Still, this policy will be enforced when an online role expires and the automatic job to bump that role is executed.
See: https://github.com/repository-service-tuf/repository-service-tuf-worker/blob/486e98cdbc16dd41c62c0ec8b473e2f94312ee9c/app.py#L139

The idea behind this new endpoint is that after you have called `PUT /api/v1/config` you can call this new endpoint `PUT /api/v1/metadata` and request a new metadata version of part or all online metadata roles.

For the future I have thought about adding a flag `force_new_version` to the `PUT /api/v1/config`  endpoint which will do the same think as this new endpoint but for all roles.

MVrachev replied:
As suggested by Ivana I added a commit that makes sure we cannot push negative expiration days.
Here is the commit: https://github.com/repository-service-tuf/repository-service-tuf-api/pull/513/commits/8a00f5bb795d3b73dc3313924c019f543302d2ef

ivanayov commented on repository_service_tuf_api/api/metadata.py:
```diff
@@ -31,6 +31,26 @@ def post(payload: metadata.MetadataPostPayload):
+    description=(
+        "Force a new version of online metadata role(s). If the roles list is "
+        "empty all roles will be updated. The new metadata version(s) will "
+        "have extended expiration which will equal to: "
```
Same about "extended" here.

MVrachev replied:
Check my comment to your questions above.

MVrachev commented:
I had to remove my latest changes where I added more requirements to our attributes as there was an issue.
I have added a new issue about this: #520.

MVrachev commented on repository_service_tuf_api/metadata.py:
```diff
@@ -85,6 +85,79 @@ def post_metadata(payload: MetadataPostPayload) -> MetadataPostResponse:
 
 
+class MetadataPutPayload(BaseModel):
+    roles: List[Roles.online_roles_values()]
```
We need to change this as we would support custom target delegations' role names.
This should be any string.

MVrachev commented on repository_service_tuf_api/config.py:
```diff
@@ -63,6 +63,8 @@ class Config:
 
 class Settings(BaseModel):
     # Only online roles can be dict keys
+    # TODO: Add requirement for int value after fixing: https://github.com/repository-service-tuf/repository-service-tuf-api/issues/377  # noqa
```
I need to remove those comments as we have removed the FAST API issue.

MVrachev commented:
@kairoaraujo ready to test together with https://github.com/repository-service-tuf/repository-service-tuf-worker/pull/435.
Support for pushing a new version of custom target delegation is there.
